### DISCUSSION TITLE
override computeLocalScopes() issue

### DISCUSSION THREAD

**Discussion by simpsoneric** [eclipse-langium/langium#869] [Q&A] [answered]
This is probably user error because I'm learning both langium and typescript.

I'm trying to utilize an override for `ScopeComputation::computeLocalScopes()`

In the base DefaultScopeComputation, it is defined as:

```javascript
DefaultScopeComputation {

    async computeLocalScopes(
                document: LangiumDocument,
                cancelToken = CancellationToken.None): Promise<PrecomputedScopes>
    {...}
}
```

And in the domainmodel example it is:

https://github.com/langium/langium/blob/938bd0da1d7b66d1560f0d3a07a5b1f4d5753099/examples/domainmodel/src/language-server/domain-model-scope.ts#L45

But if I use the same override for my custom language, I get

```
Error: Cannot find module 'vscode'
Require stack:
- /aadl-lang/node_modules/vscode-languageclient/lib/node/main.js
- /aadl-lang/out/language-server/aadl-scope.js
- /aadl-lang/out/language-server/aadl-module.js
- /aadl-lang/out/language-server/main.js
    at Function._resolveFilename (node:internal/modules/cjs/loader:987:15)
    at node:internal/modules/cjs/loader:832:27
    at Function._load (node:electron/js2c/asar_bundle:5:13343)
    at Module.require (node:internal/modules/cjs/loader:1059:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/aadl-lang/node_modules/vscode-languageclient/lib/node/main.js:25:18)
    at Module._compile (node:internal/modules/cjs/loader:1165:14)
    at Object..js (node:internal/modules/cjs/loader:1220:10)
    at Module.load (node:internal/modules/cjs/loader:1035:32)
    at node:internal/modules/cjs/loader:876:12 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/aadl-lang/node_modules/vscode-languageclient/lib/node/main.js',
    '/aadl-lang/out/language-server/aadl-scope.js',
    '/aadl-lang/out/language-server/aadl-module.js',
    '/aadl-lang/out/language-server/main.js'
  ]
}
[Error - 5:15:35 AM] The aadl server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
[Error - 5:15:35 AM] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 5:15:35 AM] aadl client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 5:15:35 AM] Restarting server failed
  Message: Pending response rejected since connection got disposed
  Code: -32097
```

This is roughly the same error as: https://github.com/langium/langium/discussions/733

Except I have grepped through my module code and there is no `import vscode`.

Instead, if I change the override to:

```javascript
override async computeLocalScopes(
    document: LangiumDocument, 
    cancelToken?: CancellationToken): Promise<PrecomputedScopes> 
    {...}
```

The extension behaves correctly.
Can anyone provide or point me to an explanation for what is going wrong with the previous override?

**Comment by msujew** [comment, marked as answer]
Hey @simpsoneric,

It looks like you're importing `vscode-languageclient` in your language server code. The package requires the `vscode` import to the available which is only the case in the extension code. Your can simply replace those imports with `vscode-languageserver`.

**Reply by simpsoneric** [reply]
Ah, yes I think the vscode auto import fix added:

```javascript
import { CancellationToken } from "vscode-languageclient";
```

I replaced it with:

```javascript
import { CancellationToken } from "vscode-languageserver";
```

and the previous errors seem to have gone away.
the below is now functional.

```javascript
    override async computeLocalScopes(document: LangiumDocument, cancelToken = CancellationToken.None): Promise<PrecomputedScopes> {
```

**Reply by spoenemann** [reply]
If you want to avoid such situations, you can separate the language server and the extension into different packages. You just need to ensure that your bundler (esbuild, webpack etc.) writes the output file into the extension package. With that setup, each package has its own set of dependencies: vscode-languageserver in the language server, vscode-languageclient in the extension.
